File: 天梯赛/L3-004. 肿瘤诊断.cpp
#include<iostream>
#include<queue>
#include<algorithm>
using namespace std;
int table[1300][130][70];//x,y,z
bool vis[1300][130][70];
struct node{
	int x,y,z;
};
int dx[]={0,0,1,0,0,-1};
int dy[]={0,1,0,0,-1,0};
int dz[]={1,0,0,-1,0,0};
int m,n,l,t;
bool judge(int x,int y,int z){
	if(x<0||y<0||z<0||x>=m||y>=n||z>=l)return false;
	if(vis[x][y][z])return false;
	if(!table[x][y][z])return false;
	return true;
}
int bfs(int x,int y,int z){
	int cnt=0;
	queue<node> q;
	vis[x][y][z]=true;
	node tmp,nex;
	tmp.x=x;tmp.y=y;tmp.z=z;
	q.push(tmp);
	while(!q.empty()){
		tmp=q.front();q.pop();
		cnt++;
		for(int i=0;i<6;++i){
			int newx=tmp.x+dx[i];
			int newy=tmp.y+dy[i];
			int newz=tmp.z+dz[i];
			if(judge(newx,newy,newz)){
				vis[newx][newy][newz]=true;
				nex.x=newx;nex.y=newy;nex.z=newz;
				q.push(nex);
			}
		}
	}
	if(cnt>=t)return cnt;
	else return 0;
}
int main(){
	int res=0;
	cin>>m>>n>>l>>t;
	for(int i=0;i<l;++i)
		for(int j=0;j<m;++j)
			for(int k=0;k<n;++k)
				cin>>table[j][k][i];
	for(int i=0;i<l;++i)
		for(int j=0;j<m;++j)
			for(int k=0;k<n;++k){
				if(!vis[j][k][i]&&table[j][k][i])
					res+=bfs(j,k,i);
			}
	printf("%d\n",res);
	return 0;
}

File: 天梯赛/java/L3-004. 肿瘤诊断.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Scanner;
import java.util.Stack;

public class Main {
	public static int table[][][]=new int[1300][130][70];
	public static boolean vis[][][]=new boolean [1300][130][70];
	public static int dx[]={0,0,1,0,0,-1};
	public static int dy[]= {0,1,0,0,-1,0};
	public static int dz[]= {1,0,0,-1,0,0};
	public static int m,n,l,t;
	public static void main(String[] args) {
		Scanner in=new Scanner(System.in);
		int res=0;
		m=in.nextInt();n=in.nextInt();l=in.nextInt();t=in.nextInt();
		for(int i=0;i<l;++i)
			for(int j=0;j<m;++j)
				for(int k=0;k<n;++k)
					table[j][k][i]=in.nextInt();
		in.close();
		for(int i=0;i<l;++i)
			for(int j=0;j<m;++j)
				for(int k=0;k<n;++k)
					if(!vis[j][k][i]&&table[j][k][i]!=0)res+=bfs(j,k,i);
		System.out.print(res);
	}

	public static boolean judge(int x,int y,int z) {
		if(x<0||y<0||z<0||x>=m||y>=n||z>=l)return false;
		if(vis[x][y][z])return false;
		if(table[x][y][z]==0)return false;
		return true;
	}
	
	public static int bfs(int x,int y,int z) {
		int cnt=0;
		Queue<node> q=new LinkedList<>();
		vis[x][y][z]=true;
		q.add(new node(x,y,z));
		while(!q.isEmpty()) {
			node tmp=q.poll();
			cnt++;
			for(int i=0;i<6;++i) {
				int nx=tmp.x+dx[i];
				int ny=tmp.y+dy[i];
				int nz=tmp.z+dz[i];
				if(judge(nx,ny,nz)) {
					vis[nx][ny][nz]=true;
					q.add(new node(nx,ny,nz));
				}
			}
		}
		if(cnt>=t)return cnt;
		return 0;
	}
}

class node{
	int x,y,z;
	node(int x,int y,int z){
		this.x=x;
		this.y=y;
		this.z=z;
	}
}